Classes/Private/Manager/Audio/OCTAudioEngine.m

```diff
@@ -0,0 +1,237 @@
+//
+//  OCTAudioEngine.m
+//  objcTox
+//
+//  Created by Chuong Vu on 5/24/15.
+//  Copyright (c) 2015 dvor. All rights reserved.
+//
+
+#import "OCTAudioEngine.h"
+@import AVFoundation;
+
+static const AudioUnitElement kInputBus = 1;
+static const AudioUnitElement kOutputBus = 0;
+
+@interface OCTAudioEngine ()
+
+@property (nonatomic, assign) AUGraph processingGraph;
+@property (nonatomic, assign) AUNode ioNode;
+@property (nonatomic, assign) AudioUnit ioUnit;
+@property (nonatomic, assign) AudioComponentDescription ioUnitDescription;
+
+@end
+
+@implementation OCTAudioEngine
+
+#pragma mark - LifeCycle
+-(instancetype)init
+{
+    self = [super init];
+    if (!self)
+        return nil;
+
+    _ioUnitDescription.componentType = kAudioUnitType_Output;
+    _ioUnitDescription.componentSubType = kAudioUnitSubType_VoiceProcessingIO;
+    _ioUnitDescription.componentManufacturer = kAudioUnitManufacturer_Apple;
+    _ioUnitDescription.componentFlags = 0;
+    _ioUnitDescription.componentFlagsMask = 0;
+
+    NewAUGraph(&_processingGraph);
+    NewAUGraph(&_processingGraph);
+
+    AUGraphAddNode(_processingGraph,
+                   &_ioUnitDescription,
+                   &_ioNode);
+
+    AUGraphOpen(_processingGraph);
+
+    AUGraphNodeInfo(_processingGraph, _ioNode, NULL, &_ioUnit);
+    return self;
+}
+
+#pragma mark - Audio Controls
+-(BOOL)startAudioFlow:(NSError **)error
+{
+    if (![self startAudioSession:error] ||
+        ![self microphoneInput:YES error:error] ||
+        ![self setUpStreamFormat:error] ||
+        ![self initializeGraph:error] ||
+        ![self startGraph:error]) {
+        return NO;
+    }
+
+    return YES;
+}
+
+-(BOOL)stopAudioFlow:(NSError **)error
+{
+    OSStatus status = AUGraphStop(_processingGraph);
+    if (status != noErr) {
+        *error = [self createErrorWithCode:status
+                               description:@"AUGraphStop"
+                             failureReason:@"Failed to stop Graph"];
```

Comment:
It will crash in case if error will be `nil`. You have to perform `if (error)` check before assigning. To DRY you can pass error to method and in it check:
```
[self fillError:error withCode:code description:@"AUGraphStop" failureReason:@"Failed to stop Graph"];

...

- (void)fillError:(NSError **)error ...
{
    if (! error) {
        return;
    }
    ...
    *error = ...
}
```

Same about all other errors.